agent/registration/csr.go

```diff
@@ -0,0 +1,95 @@
+// Copyright 2021 VMware, Inc. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package registration
+
+import (
+	"context"
+	"crypto/rand"
+	"crypto/rsa"
+	"crypto/x509"
+	"crypto/x509/pkix"
+	"encoding/pem"
+	"fmt"
+
+	certv1 "k8s.io/api/certificates/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/klog/v2"
+	"k8s.io/utils/pointer"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+const (
+	KeySize           = 2048
+	ExpirationSeconds = 86400
+	ByohCSROrg        = "byoh:hosts"
+)
+
+type ByohCSR struct {
+	K8sClient client.Client
+}
+
+func (bcsr *ByohCSR) CreateCSR(hostname, namespace string) (*rsa.PrivateKey, error) {
+	ctx := context.TODO()
+	privKey := &rsa.PrivateKey{}
+	byoCSR := &certv1.CertificateSigningRequest{}
+	err := bcsr.K8sClient.Get(ctx, types.NamespacedName{Name: hostname, Namespace: namespace}, byoCSR)
+	if err != nil {
+		if !apierrors.IsNotFound(err) {
+			klog.Errorf("error getting csr %s in namespace %s, err=%v", hostname, namespace, err)
+			return nil, err
```

Comment:
I'm trying to understand the difference between `return nil` here and `return privKey` on line# 55
Can we also `return privKey` here? That's going to be an empty `PrivateKey{}`

Reply:
When the flow reaches line #55, it means there are no errors, hence we are returning privateKey. In other instances, an error is thrown and IMHO I do not find any value returning empty value over return nil.